src/roswire/proxy/roslaunch/reader.py

```diff
@@ -23,10 +25,33 @@
 from ...name import (global_name, name_is_global, name_is_private, namespace,
                      namespace_join)
 
+if typing.TYPE_CHECKING:
+    from ...app import AppInstance
+
+
+@attr.s(auto_attribs=True)
```

Comment:
#408

Reply:
Using `attrs` turns this into a partial implementation and uses inheritance to implement the subclasses. We're better off using `abc.ABC` and keeping this as a pure interface.